android/README.md

```diff
@@ -16,73 +25,79 @@ Build the demo module to see it in action, or just call this in `Activity`'s `on
+    
+There are two options for submitting envoy urls:
+    
+ - submit(context: Context, urls: List<String>)
+ - submit(context: Context, urls: List<String>, directUrl: String)
+    
+If the optional directUrl parameter is included, Envoy will attempt to connect to that url directly first.  This can be included to avoid using proxy resources when the target domain is not blocked.
+
+## Basic envoy integration
+
+```kotlin
+    private val mBroadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
+        override fun onReceive(context: Context?, intent: Intent?) {
+            if (intent != null && context != null) {
+                if (intent.action == org.greatfire.envoy.BROADCAST_URL_VALIDATION_SUCCEEDED) {
+                    val validUrls = intent.getStringArrayListExtra(org.greatfire.envoy.EXTENDED_DATA_VALID_URLS)
+                    if (validUrls != null && !validUrls.isEmpty()) {
+                        val envoyUrl = validUrls[0]
+                        CronetNetworking.initializeCronetEngine(context, envoyUrl)
+                    } else {
+                        // received empty list
+                    }
+                } else if (intent.action == org.greatfire.envoy.BROADCAST_URL_VALIDATION_FAILED) {
+                    val invalidUrls = intent.getStringArrayListExtra(org.greatfire.envoy.EXTENDED_DATA_INVALID_URLS)
+                    if (invalidUrls != null && !invalidUrls.isEmpty()) {
+                        // handle error state
+                    } else {
+                        // received empty list
+                    }
+                } else {
+                    // received unexpected intent
                 }
-            } else if (intent.action == BROADCAST_URL_VALIDATION_FAILED) {
-                final List<String> invalidUrls = intent.getStringArrayListExtra(org.greatfire.envoy.NetworkIntentServiceKt..EXTENDED_DATA_INVALID_URLS);
-                Log.i("BroadcastReceiver", "Received invalid urls: " + TextUtils.join(", ", validUrls));
-            } else if (intent.action == ShadowsocksService.SHADOWSOCKS_SERVICE_BROADCAST) {
-                final int shadowsocksResult = intent.getIntExtra(ShadowsocksService.SHADOWSOCKS_SERVICE_RESULT, 0);
-                // check service result as described above
-                // consider submitting urls only after the shadowsocks service has had a chance to start
+            } else {
+                // received null intent or context
             }
         }
     }
-};
+    
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        LocalBroadcastManager.getInstance(this).registerReceiver(mBroadcastReceiver, IntentFilter().apply {
+            addAction(org.greatfire.envoy.BROADCAST_URL_VALIDATION_SUCCEEDED)
+            addAction(org.greatfire.envoy.BROADCAST_URL_VALIDATION_FAILED)
+        })
+    
+        val listOfUrls = mutableListOf<String>()
+        listOfUrls.add(foo)
```

Comment:
um...

Reply:
i'll do some edits, but this is just an example so i'm not going to use actual urls or config parameters.